main.cpp

```diff
@@ -17,41 +17,79 @@ Purpose:  This project will teach you about variadic templates and recursive tem
 Make the following program work, which makes use of Variadic templates and Recursion
  */
 
-#include <iostream>
-#include <string>
-#include <typeinfo>
-
-struct Point
-{
-    Point(float _x, float _y) : x(_x), y(_y) { }
-    Point& multiply(float m)
-    {
-        x *= m;
-        y *= m;
-        return *this;
-    }
-    std::string toString() const
-    {
-        std::string str;
-        str +="Point { x: ";
-        str += std::to_string(x);
-        str += ", y: ";
-        str += std::to_string(y);
-        str += " }";
-        return str;
-    }
-private:
-    float x{0}, y{0};
-};
-
-template<typename Type>
-struct Wrapper
-{
-    Wrapper(Type&& t) : val(std::move(t)) 
-    { 
-        std::cout << "Wrapper(" << typeid(val).name() << ")" << std::endl; 
-    }
-};
+#include <iostream>
+#include <string>
+#include <typeinfo>
+#include "typename.h"
+
+struct Point
+{
+    Point(float _x, float _y) : x(_x), y(_y) { }
+    Point& multiply(float m)
+    {
+        x *= m;
+        y *= m;
+        return *this;
+    }
+    std::string toString() const
+    {
+        std::string str;
+        str +="Point { x: ";
+        str += std::to_string(x);
+        str += ", y: ";
+        str += std::to_string(y);
+        str += " }";
+        return str;
+    }
+private:
+    float x{0}, y{0};
+};
+
+template<typename Type>
+struct Wrapper
+{
+    Wrapper(Type&& t) : val(std::move(t))
+    { 
+        std::cout << "Wrapper(" << type_name<decltype(val)>() << ")" << std::endl;
+    }
+
+    void print()
+    {
+        std::cout << "Wrapper::print(" << val << ")" << std::endl;
+    }
+
+private:
+    Type val;
+};
+
+template<>
+struct Wrapper<Point>
+{
+    Wrapper(Point&& p) : point(std::move(p))
+    { 
+        std::cout << "Wrapper(" << type_name<decltype(point)>() << ")" << std::endl;
+    }
+
+    void print()
+    {
+        std::cout << "Wrapper::print(" << point.toString() << ")" << std::endl;
+    }
+private:
+    Point point;
+};
```

Comment:
Project 9.1 Challenge 1:
There is a lot of repetition between the template class `Wrapper`, and the template specialization for `Wrapper<Point>`
Can you figure out how to eliminate the duplicate constructor?
hint: you only have to _explicitly specialize_ that which is unique.
You can specialize member functions without specializing the entire class :wink: